### PR TITLE
Compilation on GCC 11

### PR DESCRIPTION
**Pull request by nwu63**
## Purpose
I fixed a few things in the source code such that it can be compiled by gcc 11.
- The main thing to fix was rank mismatch between 1D array and scalar. To fix this, I converted them to singleton arrays and differentiated with Tapenade
- I deleted one file in `outputReverse` that I believe is unused, since the same code exists in `kd_tree` which is specifically hand-differentiated.

## Type of change

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [x] Maintenance update
- [ ] Other (please describe)

## Testing
Both real and complex tests now pass.

## Checklist

- [x] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
- [x] I have run unit and regression tests which pass locally with my changes
- [ ] I have added new tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation

### CODE REVIEW

File: src/warp/outputReverse/getelementprops_b.f90 [hunk 1]
```diff
@@ -22,15 +22,16 @@ SUBROUTINE GETELEMENTPROPS_B(points, pointsb, npts, area, areab, normal&
...
+   REAL(kind=realtype), DIMENSION(1) :: crossnormb
    INTRINSIC SQRT
    REAL(kind=realtype), DIMENSION(3) :: tmp
    INTEGER :: branch
    REAL(kind=realtype) :: temp0
+   REAL(kind=realtype) :: tempb1
    REAL(kind=realtype) :: tempb0
    REAL(kind=realtype) :: tmpb(3)
-   REAL(kind=realtype) :: tempb
+   REAL(kind=realtype) :: tempb(3)
```
Comment (bernardopacini) [inline src/warp/outputReverse/getelementprops_b.f90:34]: How did this become length 3?

File: src/modules/kd_tree.F90 [hunk 2]
```diff
@@ -1104,7 +1104,7 @@ SUBROUTINE COMPUTENODALPROPERTIES_B(tp, initialpoint)
           CALL PUSHINTEGER4(kk-1)
           CALL PUSHREAL8ARRAY(facenormal, realtype*3/8)
           CALL GETELEMENTPROPS(points, npts, facearea, facenormal)
-          CALL PUSHREAL8ARRAY(da, realtype/8)
+          CALL PUSHREAL8(da, realtype/8)
```
Comment (bernardopacini) [inline src/modules/kd_tree.F90:1107]: Is this done manually or with Tapenade?

**Review by bernardopacini — COMMENTED** [review]
Overall this looks good to me. Do you know of possible edge-cases where these changes would compile with GCC 11 but not work at runtime?

File: src/warp/outputReverse/getelementprops_b.f90 (hunk 1 above)
Comment (marcomangano) [reply src/warp/outputReverse/getelementprops_b.f90:34]: I noticed that as well, but his looks like some tapenade trickery, not sure if it makes sense to worry about it

Comment (nwu63) [reply src/warp/outputReverse/getelementprops_b.f90:34]: Yeah I have no idea, but this was auto-generated by tapenade so who knows :shrug:

File: src/modules/kd_tree.F90 (hunk 2 above)
Comment (nwu63) [reply src/modules/kd_tree.F90:1107]: This was done manually, since this is a hand-differentiated routine (I think it was actually differentiated by Tapenade and modified by hand extensively), but in any case the file is not "controlled" by Tapenade I believe.

**Comment by nwu63** [comment]
> Do you know of possible edge-cases where these changes would compile with GCC 11 but not work at runtime?

Not that I know of, and we do not test against GCC 11 anyways (I think we are on 8/9 in the testing images) so I think this is fine. I am running GCC 11 locally which is the motivation behind this PR.

File: src/warp/outputReverse/getelementprops_b.f90 (hunk 1 above)
Comment (bernardopacini) [reply src/warp/outputReverse/getelementprops_b.f90:34]: Looking at this again, it seems that Tapenade has rewritten a large chunk of the section that uses that variable. Not an issue, I am sure it is referencing it properly.

**Comment by nwu63** [comment]
@bernardopacini or @anily could you take a look at this PR?

**Review by bernardopacini — APPROVED** [review]

**Comment by anilyil** [comment]
Sorry for the late reply, before we merge this, can someone add a comment in the code about how that routine that looks like its differentiated by tapenade is actually hand differentiated (or at least modified by hand etc)? Also, was there no way around using scalars instead of a single entry arrays to get this working?

**Comment by anilyil** [comment]
Would that flag hide any other inconsistencies when adding new code?

**Comment by bernardopacini** [comment]
> Would that flag hide any other inconsistencies when adding new code?

It doesn't hide them, it just converts the mismatch errors into warnings. It does make it easier to mistakenly pass a scalar for a 1 element array (what we are trying to allow here), or something like that, but you will at least be warned on compile time.

**Comment by anilyil** [comment]
Okay, that makes sense. I don't want to delay this PR any further for that "nice to have". This is definitely code gymnastics that we probably don't want, but the code probably already has tons of stuff like this. Ultimately, it's up to you. I think there is value in fixing this w/o arrays, and I am surprised there is no other way besides that flag. 

The only thing that would be good to add is a comment in the code that highlights the hand-differentiated or hand-modified tapenade code.

**Comment by bernardopacini** [comment]
I agree. The best solution here would be to have the team developing Tapenade update their end so that it all works properly with GCC 10+. But, I don't think this is realistic in the time frame we are working for. I am fine with either option, I have gone with the flag option in the past simply because it does not require gymnastics, as you said.

If we stick with this, the changes are good by me. But I do personally lean towards the flag option just so that we avoid possible modifications like this in the future, whenever an incompatibility arises.

**Comment by nwu63** [comment]
- The comments just before `computenodalproperties_b` says
```
  ! This originally came from tapenade, but has to extensively
  ! manually modified for it work iwthout having an enitrely new
  ! kd_tree_b.
```
- I forget why `crossNorm` had to be defined as a vector, but I will try it again this afternoon.

**Comment by anilyil** [comment]
Sorry I did not see that comment. Based on the discussion here I thought there was not such an explanation in the code so that is fine. I agree with avoiding the compiler flag.

**Comment by nwu63** [comment]
Okay I now remember, the rank mismatch issue has to do with TAPENADE using these ARRAY functions for some reason even if the arguments are scalar. In the future maybe if we switch to a newer version of TAPENADE it would be smarter, but I had trouble getting 3.16 working. I think we should just merge this.

**Comment by bernardopacini** [comment]
Quick note: 3.16 has the same issue so that unfortunately won't fix it. Merging this sounds good to me.

**Comment by nwu63** [comment]
> Quick note: 3.16 has the same issue so that unfortunately won't fix it. Merging this sounds good to me.

Oh right, because this is stuff in the ADFirstAidKit?

**Comment by bernardopacini** [comment]
> > Quick note: 3.16 has the same issue so that unfortunately won't fix it. Merging this sounds good to me.
> 
> Oh right, because this is stuff in the ADFirstAidKit?

Yes, exactly.

**Review by anilyil — APPROVED** [review]

**Comment by anilyil** [comment]
Sounds good. I approved but I am also not in this list so it probably does not count. Merging is fine for me.

**Comment by nwu63** [comment]
I will merge since we have three approvals.
